storage/mongo/storage.go

```diff
@@ -0,0 +1,251 @@
+		BID:    bson.NewObjectId(),
+	}
+
+	ops := []txn.Op{{
+		C:      user.TableName(),
+		Id:     user.ID,
+		Update: bson.M{"$set": user},
+	}, {
+		C:      token.TableName(),
+		Id:     token.BID,
+		Insert: token,
+	}}
+
+	if err := runner.Run(ops, bson.NewObjectId(), nil); err != nil {
+		return nil, errors.Wrap(err, "error granting authenticated user")
+	}
+
+	return token, nil
+}
+
+func (conn *Connection) GrantRefreshTokenSwap(user *models.User, token *models.RefreshToken) (*models.RefreshToken, error) {
+	runner := conn.newTxRunner()
+
+	token.Revoked = true
+
+	newToken := &models.RefreshToken{
+		User:   *user,
+		UserID: user.ID,
+		Token:  crypto.SecureToken(),
+		BID:    bson.NewObjectId(),
+	}
+
+	ops := []txn.Op{{
+		C:      token.TableName(),
+		Id:     token.BID,
+		Update: bson.M{"$set": token},
+	}, {
+		C:      token.TableName(),
+		Id:     newToken.BID,
+		Insert: newToken,
+	}}
+
+	if err := runner.Run(ops, bson.NewObjectId(), nil); err != nil {
+		return nil, errors.Wrap(err, "error granting authenticated user")
+	}
+
+	return token, nil
+}
+
+func (conn *Connection) IsDuplicatedEmail(email, id string) (bool, error) {
+	_, err := conn.findUser(bson.M{"email": email, "_id": bson.M{"$ne": id}})
+	if err != nil {
+		if err == mgo.ErrNotFound {
+			return false, nil
+		} else {
+			return false, errors.Wrap(err, "error checking duplicated email")
+		}
+	}
+
+	return true, nil
+}
+
+func (conn *Connection) Logout(id interface{}) {
+	c := conn.db.C(models.RefreshToken{}.TableName())
+	c.RemoveAll(bson.M{"user_id": id})
+}
+
+func (conn *Connection) RevokeToken(token *models.RefreshToken) error {
+	token.Revoked = true
+
+	c := conn.db.C(token.TableName())
+	if err := c.Update(bson.M{"_id": token.BID}, bson.M{"$set": token}); err != nil {
+		return errors.Wrap(err, "error revoking refresh token")
+	}
+	return nil
+}
+
+func (conn *Connection) RollbackRefreshTokenSwap(newToken, oldToken *models.RefreshToken) error {
+	runner := conn.newTxRunner()
+
+	newToken.Revoked = true
+	oldToken.Revoked = false
+
+	ops := []txn.Op{{
+		C:      newToken.TableName(),
+		Id:     newToken.BID,
+		Update: bson.M{"$set": newToken},
+	}, {
+		C:      oldToken.TableName(),
+		Id:     oldToken.BID,
+		Update: bson.M{"$set": oldToken},
+	}}
+
+	if err := runner.Run(ops, bson.NewObjectId(), nil); err != nil {
+		return errors.Wrap(err, "error granting authenticated user")
+	}
+
+	return nil
+}
+
+func (conn *Connection) UpdateUser(user *models.User) error {
+	c := conn.db.C(user.TableName())
+	if err := c.Update(bson.M{"_id": user.ID}, bson.M{"$set": user}); err != nil {
+		return errors.Wrap(err, "Error updating user record")
+	}
+	return nil
+}
+
+func (conn *Connection) newTxRunner() *txn.Runner {
+	return txn.NewRunner(conn.db.C("auth_transactions"))
+}
+
+// Connect will connect to that storage engine
+func Connect(config *conf.Configuration) (*Connection, error) {
+	mgo.SetLogger(logger{logrus.WithField("db-connection", "mongo")})
```

Comment:
omg - I haven't seen this before. I always have a `logrus.NewEntry(logrus.StandardLogger())`. This is sooo much easier.